sumpy/fmm.py

```diff
@@ -353,27 +354,14 @@ def __init__(self, tree_indep, traversal, dtype, fmm_level_to_order,
         self.extra_kwargs = source_extra_kwargs.copy()
         self.extra_kwargs.update(self.kernel_extra_kwargs)
 
-        if base_kernel.is_translation_invariant:
-            if translation_classes_data is None:
-                from warnings import warn
-                if self.tree_indep.use_fft_for_m2l:
-                    raise NotImplementedError(
-                         "FFT based List 2 (multipole-to-local) translations "
-                         "without translation_classes_data argument is not "
-                         "implemented. Supply a translation_classes_data argument "
-                         "to the wrangler for optimized List 2.")
-                else:
-                    warn(
-                         "List 2 (multipole-to-local) translations will be "
-                         "unoptimized. Supply a translation_classes_data argument "
-                         "to the wrangler for optimized List 2.",
-                         SumpyTranslationClassesDataNotSuppliedWarning,
-                         stacklevel=2)
-                self.supports_translation_classes = False
-            else:
-                self.supports_translation_classes = True
-        else:
+        if _disable_translation_classes or not base_kernel.is_translation_invariant:
             self.supports_translation_classes = False
+        else:
+            if translation_classes_data is None:
+                with cl.CommandQueue(self.tree_indep.cl_context) as queue:
+                    translation_classes_data = SumpyTranslationClassesData(queue,
+                        self.traversal)
```

Comment:
Why does this work? Doesn't the queue get destroyed when it exits the context manager leaving `SumpyTranslationClassesData` holding an invalid reference?

Reply:
Yes, the queue completes and gets destroyed, but the data itself doesn't get destroyed. i.e. the data has `queue=None`. When a new queue needs this data, they are accessed with `.with_queue(queue)`

Reply:
https://github.com/inducer/sumpy/blob/a2c830abed936e42424aad025b6d44bd82bb7750/sumpy/fmm.py#L244-L249
Hm, but the constructor doesn't do any work. The actual work is being done when calling some of those memoized methods below, which should be after the context manager exists.

Reply:
The tests don't seem to crash and burn, so this is probably just some misunderstanding on my side. Hm..

Reply:
Hmm, you are right. I guess since the constructor keeps a reference to the queue, it doesn't get destroyed.

Reply:
https://github.com/inducer/pyopencl/blob/175060ab392e40866dc1291eda13fec724ac7acc/pyopencl/__init__.py#L707-L708

Yeah, you're right! `__exit__` doesn't destroy it, it just calls finish, so should be ok!

Reply:
Good catch, @alexfikl! I was under the mistaken assumption that the work got done by the constructor.

The context manager does not destroy the queue:

https://github.com/inducer/pyopencl/blob/175060ab392e40866dc1291eda13fec724ac7acc/pyopencl/__init__.py#L704-L708=

But continuing to hold a reference to the queue past `__exit__` is definitely not the intended use of the interface.

(Fun fact: Neither CL nor PyOpenCL allows destroying a queue, only releasing a reference to it.)

Reply:
I'll try to come up with a way to make PyOpenCL warn if a reference to the queue continues to be held past `__exit__`.

Reply:
That would be really cool!

Reply:
I changed the implementation to build the data objects eagerly instead of lazily.

Reply:
Can't speak for the eager vs lazy, but the queue stuff makes sense to me now :+1:

Reply:
https://github.com/inducer/pyopencl/pull/561